src/main/java/de/tum/cit/aet/artemis/atlas/service/CompetencyExpertToolsService.java

```diff
@@ -0,0 +1,459 @@
+     * @param viewOnly     optional flag for view-only mode
+     * @return Simple confirmation message for the LLM to use in its response
+     */
+    @Tool(description = "Preview one or multiple competencies before creating/updating. SINGLE: pass [{comp}]. BATCH: pass [{comp1}, {comp2}, {comp3}]. CRITICAL: For batch operations, pass ALL competencies in ONE call, not multiple separate calls.")
+    public String previewCompetencies(@ToolParam(description = "the Course ID from the CONTEXT section") Long courseID,
+            @ToolParam(description = "list of competency operations to preview") List<CompetencyOperation> competencies,
+            @ToolParam(description = "optional: set to true for view-only mode (no action buttons)", required = false) Boolean viewOnly) {
+        if (competencies == null || competencies.isEmpty()) {
+            return "Error: No competencies provided for preview.";
+        }
+
+        // Validate all competencies before processing
+        for (CompetencyOperation comp : competencies) {
+            String validationError = validateCompetencyOperation(comp);
+            if (validationError != null) {
+                return "Error: " + validationError;
+            }
+        }
+
+        // Store preview data in ThreadLocal for deterministic extraction by AtlasAgentService
+        List<CompetencyPreviewDTO> previewResponses = competencies.stream().map(comp -> new CompetencyPreviewDTO(comp.getTitle(), comp.getDescription(),
+                comp.getTaxonomy().toString(), getTaxonomyIcon(comp.getTaxonomy()), comp.getCompetencyId(), viewOnly)).toList();
+
+        currentPreviews.set(previewResponses);
+
+        // Cache the competency operation data for refinement operations
+        // This enables deterministic modifications (e.g., changing taxonomy while preserving title/description)
+        String sessionId = currentSessionId.get();
+        if (sessionId != null && !Boolean.TRUE.equals(viewOnly)) {
+            // Only cache if not in view-only mode (view-only is for browsing, not editing)
+            atlasAgentService.cachePendingCompetencyOperations(sessionId, new ArrayList<>(competencies));
+        }
+
+        // Return simple confirmation message that the LLM can use naturally in its response
+        // The actual preview data will be extracted from ThreadLocal by AtlasAgentService
+        if (competencies.size() == 1) {
+            return "Preview generated successfully for 1 competency.";
+        }
+        else {
+            return "Preview generated successfully for " + competencies.size() + " competencies.";
+        }
+    }
+
+    /**
+     * Maps a CompetencyTaxonomy to its corresponding icon name.
+     * This mapping is critical for client-side display and must remain stable.
+     *
+     * @param taxonomy the competency taxonomy
+     * @return the corresponding icon name for FontAwesome
+     */
+    private String getTaxonomyIcon(CompetencyTaxonomy taxonomy) {
```

Comment:
This ties the server implementation to the client application. We need to decouple this.